### PR TITLE
Feature/raterule converter example 3.7

### PR DESCRIPTION
**Pull request by alessandrofelder**
## Description
This is an initial implementation of parts of Algorithm 3.1 described in Fages et al, Theoretical Computer Science, 2015.
The parts are sufficient to handle example 3.7 in the publication, which is implemented as a test here too.

## Motivation and Context
It's a step on the way to get the `SBMLRateRuleConverter` working.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Change in documentation

## Checklist:

- [x] I have updated all documentation necessary.
- [x] I have checked spelling in (new) comments.

## Testing
- [X] Testing is done automatically and codecov shows test coverage
- [ ] This cannot be tested automatically

### CODE REVIEW

**Comment by alessandrofelder** [comment]
@skeating would you mind having a look a this (I can't request your review because it's on my fork - sorry🙄 )?
I'm keen to get your thoughts on these changes before I implement the other parts (they will be somewhat analogous).
The branch I'm merging into (feature/raterule_converter) is a copy of your `sk-old-branch` with some updated unrelated files.

**Comment by alessandrofelder** [comment]
Tests possibly failing due to old C++ standard?

File: src/sbml/conversion/SBMLRateRuleConverter.cpp [hunk 1]
```diff
@@ -667,6 +666,86 @@ SBMLRateRuleConverter::populateODEinfo()
...
+              // add raterule defining dz/dz = -dxdt
+              RateRule* raterule = model->createRateRule();
+              raterule->setVariable(zName);
+              ASTNode* math = new ASTNode(ASTNodeType_t::AST_TIMES);
+              ASTNode* minus1 = new ASTNode(ASTNodeType_t::AST_REAL);
+              minus1->setValue(-1.0);
+              ASTNode* dxdt = odeRHS->deepCopy();
+              math->addChild(minus1);
+              math->addChild(dxdt);
+              raterule->setMath(math); // will math get deleted when raterule goes out of scope?
```
Comment (skeating) [inline src/sbml/conversion/SBMLRateRuleConverter.cpp]: raterule shouldnt go out of scope as creating within the model means it is copied there

File: src/sbml/conversion/SBMLRateRuleConverter.cpp [hunk 2]
```diff
@@ -667,6 +666,86 @@ SBMLRateRuleConverter::populateODEinfo()
...
+              delete minus1;
+              // TODO do z, math need deleting? Alleviate general memory leak worries?
```
Comment (skeating) [inline src/sbml/conversion/SBMLRateRuleConverter.cpp]: yes setMath creates a deepcopy so yes math here will leak

File: src/sbml/conversion/SBMLRateRuleConverter.cpp [hunk 3]
```diff
@@ -813,7 +893,133 @@ SBMLRateRuleConverter::reconstructModel()
...
+}
 
+bool SBMLRateRuleConverter::isVariableSpecies(ASTNode* node, Model* model)
+{
+    if (!node->isName()) // some nodes, like * operators, don't seem to have a name in the first place
+        return false;
+    Species* species = model->getSpecies(node->getName());
+    Parameter* parameter = model->getParameter(node->getName()); // some species in rate rules may be defined as variable parameters
+    bool isSpecies = (species != NULL || parameter != NULL);
+    bool isVariable = !node->isConstant();
```
Comment (skeating) [inline src/sbml/conversion/SBMLRateRuleConverter.cpp]: the node being constant is not a clear indicator
you will need to check whether the species or parameter isConstant = false

File: src/sbml/conversion/SBMLRateRuleConverter.cpp [hunk 4]
```diff
@@ -813,7 +893,133 @@ SBMLRateRuleConverter::reconstructModel()
...
+    return isVariable && isSpecies;
+}
+
+bool SBMLRateRuleConverter::isNumericalConstantOrParameter(ASTNode* node, Model* model)
+{
+    if (!node->isName()) // some nodes, like * operators, don't seem to have a name in the first place
+        return false;
+    Parameter* parameter = model->getParameter(node->getName());
+    bool isParameter = (parameter != NULL);
```
Comment (skeating) [inline src/sbml/conversion/SBMLRateRuleConverter.cpp]: do you need to know if the parameter is constant ??

**Review by skeating — APPROVED** [review]
couple of comments that really answers to you comments
🎉 looks great

File: src/sbml/conversion/SBMLRateRuleConverter.cpp (hunk 4 above)
Comment (alessandrofelder) [reply src/sbml/conversion/SBMLRateRuleConverter.cpp]: IIUC, yes.

File: src/sbml/conversion/SBMLRateRuleConverter.cpp (hunk 2 above)
Comment (alessandrofelder) [reply src/sbml/conversion/SBMLRateRuleConverter.cpp]: Adding
```c++
delete math
```
here causes a read access violation in the ASTNode destructor 🤔

Comment (alessandrofelder) [reply src/sbml/conversion/SBMLRateRuleConverter.cpp]: The reason for this was that I was deleting `math`'s children before deleting `math` itself.

File: src/sbml/conversion/SBMLRateRuleConverter.cpp (hunk 3 above)
Comment (alessandrofelder) [reply src/sbml/conversion/SBMLRateRuleConverter.cpp]: fixed.

**Comment by alessandrofelder** [comment]
Merging despite tests not passing on CI, as implementation incomplete anyway and relevant test passes locally for me. Will worry about CI when PR to core repo from my fork.

**Pull request merged**